File: elixir/euler/euler_problem17.exs
defmodule NumberLetterCounts do
  @moduledoc """
  https://projecteuler.net/problem=17
  """
  @num_word_map %{
    0 => "",
    1 => "one",
    2 => "two",
    3 => "three",
    4 => "four",
    5 => "five",
    6 => "six",
    7 => "seven",
    8 => "eight",
    9 => "nine",
    10 => "ten",
    11 => "eleven",
    12 => "twelve",
    13 => "thirteen",
    15 => "fifteen",
    18 => "eighteen",
    20 => "twenty",
    30 => "thirty",
    40 => "forty",
    50 => "fifty",
    80 => "eighty"
  }

  def get_from_map(num) do
    case @num_word_map |> Map.fetch(num) do
      {:ok, value} -> value
      :error -> nil
    end
  end

  
  def num2word(num) when num <= 10, do: get_from_map(num)
  def num2word(num) when num <= 20 do
    v = get_from_map(num)
    case v do
      nil -> (num |> rem(10) |> num2word()) <> "teen"
      _ -> v
    end
  end

  def num2word(num) when num < 100 do
    v = get_from_map(num)
    case v do
      nil ->
	case num |> rem(10) do
	  0 -> (num |> div(10) |> num2word()) <> "ty"
	  _ -> (((num |> div(10)) * 10) |> num2word()) <> "-" <> (num |> rem(10) |> num2word())
	end
      _ -> v
    end
  end

  def num2word(num) when num < 1000 do
    case num |> rem(100) do
      0 -> (num |> div(100) |> num2word()) <> " hundred"
      _ -> (((num |> div(100)) * 100) |> num2word()) <> " and " <> (num |> rem(100) |> num2word())
    end
  end

  def num2word(num) when num == 1000, do: "one thousand"

  def letter_count(word) do
    word |> String.replace("-", "")
    |> String.replace(" ", "")
    |> String.length()
  end


  def solution() do
    1..1000 |> Enum.map(fn x -> num2word(x) end)
    |> Enum.map(fn x -> letter_count(x) end)
    |> Enum.sum()
  end



end

File: euler/euler17/euler_problem17.c
/*
 * =====================================================================================
 *
 *       Filename:  euler_problem17.c
 *
 *    Description:  欧拉计划17题
 *
 *        Version:  1.0
 *        Created:  2018-07-27
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  wanqiang
 *
 * =====================================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

char *nums[9] = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
char *nums2[19] = {"one",     "two",       "three",    "four",     "five",
                   "six",     "seven",     "eight",    "nine",     "ten",
                   "eleven",  "twelve",    "thirteen", "fourteen", "fifteen",
                   "sixteen", "seventeen", "eighteen", "nineteen"};
char *nums3[8] = {"twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};

char *join(char *s1, char *s2) {
        char *result = malloc(strlen(s1) + strlen(s2) + 1); 
        if (result == NULL)
                exit(1);

        strcpy(result, s1);
        strcat(result, s2);

        return result;
}

void integer_to_words(int num, char **words) {
        char *ts, *hs, *rs;
        int t = num / 1000;
        if (t > 0) {
                ts = join(nums[t - 1], "thousand");
        } else {
                ts = "";
        }
        int h = num % 1000 / 100;
        if (h > 0) {
                hs = join(nums[h - 1], "hundred");
        } else {
                hs = "";
        }
        *words = join(ts, hs);
        int r = num % 100;
        if (r < 20) {
                if (r == 0)
                        rs = "";
                else
                        rs = nums2[r - 1];
        } else {
                if (num % 10 == 0)
                        rs = nums3[num % 100 / 10 - 2];
                else
                        rs = join(nums3[num % 100 / 10 - 2], nums[r % 10 - 1]);
        }
        if (r > 0 && strlen(*words) > 0)
                *words = join(*words, "and");
        *words = join(*words, rs);
}

int main() {
        clock_t begin = clock();

        int l = 0;
        for (int i = 1; i <= 1000; i++) {
                char *words = "";
                integer_to_words(i, &words);
                /* printf("%d ==> %s\n", i, words); */
                l += strlen(words);
        }
        clock_t end = clock();
        double timespent = end - begin;
        printf("result => %d, time use => %lfs\n", l, (double)(timespent / CLOCKS_PER_SEC));
        return 0;
}
